### PR TITLE
Shadow Sect

### PR DESCRIPTION
**Pull request by ro5490**
## About The Pull Request

A new sect dedicated to the expansion of darkness and shadows.

Gain Favor by putting the Obelisks in dark places, or feeding the altar flashlights.
Obelisks will automatically accrue favor, and is the main source of favor.
You start with 100 favor so you can summon your first Obelisk.

**You can only use this sect if you're already a Shadow Person**
**Antag Chaplain has access to a new shadow race mutator, role restricted antag item purchasable via uplink**

Rites - 
"Obelisk Manifestation" - 100 Favor - 45 Seconds
- Summons a Shadow Obelisk to help spread darkness.

"Shadow Expansion" - 175 Favor - 40 Seconds
- Increases the range of the darkness spread from the altar and Obelisks

"Shadowperson Conversion" - 1200 Favor - 30 Seconds
- Self explanitory

"Grant Sight" - 1000 Favor - 30 Seconds
- Gives someone shadow eyes, so they have night vision (if they dont want a full race conversion)

"Shadow Blessing" - 8000 favor - 60 seconds
- Blesses a target with anti-magic properties

...

Most of the original code is by DingoDongler and their Dark City sect

## Why It's Good For The Game

More sects good
Halloween shadow people sect, its hard for them to live on-station with all the lights

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image

image

image
image
image
image
image
image

favorgain
image
image
image

(this was cut and edited so it wouldnt be forever, but essentially shows the spread range)
spread

</details>

## Changelog
:cl:
add: Shadow Sect - Expand the darkness across the station, and convert people to the dark.
add: Shadow Obelisk for the sect that eminates dark light
add: Shadow-Person race change rite on the altar, for the shadow sect
add: New Altar sprite
add: Shadow person mutator autoinjector - Antag item bought from uplink for 5tc, Chaplain Only
/:cl:

### CODE REVIEW

**Comment by EvilDragonfiend** [comment]
🔑 I am sorry but I should put this.
While having more sects is cool, this sect gives a very shittery reason why chaplain would darken the whole station even if the space law restricts them to contain it in the chapel.
I wouldn't allow this to be in the game because of the general player behaviour. I will lift my key if you can make chaps can access this sect by unusual ways (i.e. being antag chaps, a chaplain being shadowpeople already, shadow sect unlocking tome from somewhere, admin interaction, or and so on)

**Comment by ro5490** [comment]
> 🔑 I am sorry but I should put this. While having more sects is cool, this sect gives a very shittery reason why chaplain would darken the whole station even if the space law restricts them to contain it in the chapel. I wouldn't allow this to be in the game because of the general player behaviour. I will lift my key if you can make chaps can access this sect by unusual ways (i.e. being antag chaps, a chaplain being shadowpeople already, shadow sect unlocking tome from somewhere, admin interaction, or and so on) so that people can tolerate that as it won't be introduced often.

Totally understandable, some others have already raised concerns about this so I'm implementing two main things here (excluding the two new rites that are in-progress)
This, for obvious reasons
image
 and, if I can figure out how exactly, I'm planning on having a % chance of the sect appering in the list at all, but, gotta figure out how to do that first lmao

**Comment by ro5490** [comment]
Currently just building in a if species is shadow, then unlock sect
but brain hurty

**Comment by EvilDragonfiend** [comment]
> Currently just building in a if species is shadow, then unlock sect but brain hurty

# TGUI part

image

# generate_available_sects()

image

there is `GLOB.religious_sect`. Replace for each loop in this proc using the glob, and adds a new proc to each sect datum `/proc/is_available()`

```dm
/datum/religion_sect/proc/is_available(mob/user)
// put don't call parent thing here
    return TRUE // basically all available

/datum/religion_sect/shadow/is_available(mob/user)
    if(!isshadowling(user))
        return FALSE
    return TRUE
````

**Comment by ro5490** [comment]
> > Currently just building in a if species is shadow, then unlock sect but brain hurty
> 
> # TGUI part
> 
> # generate_available_sects()
>  
> there is `GLOB.religious_sect`. Replace for each loop in this proc using the glob, and adds a new proc to each sect datum `/proc/is_available()`
> 
> ```dm
> /datum/religion_sect/proc/is_available(mob/user)
> // put don't call parent thing here
>     return TRUE // basically all available
> 
> /datum/religion_sect/shadow/is_available(mob/user)
>     if(!isshadowling(user))
>         return FALSE
>     return TRUE
> ```

So the issue is, that `easy_access_sect` gets set, and `generate_avaliable_sects()` doesnt get called, so I've been implementing a different route that should allow for "Requirements" per sect to actually unlock them-

```
/// Select the sect, called from [/datum/component/religious_tool/proc/AttemptActions]
/datum/component/religious_tool/proc/select_sect(mob/living/user, path)
	if(!ispath(text2path(path), /datum/religion_sect))
		message_admins("[ADMIN_LOOKUPFLW(usr)] has tried to spawn an item when selecting a sect.")
		return
	if(user.mind.holy_role != HOLY_ROLE_HIGHPRIEST)
		to_chat(user, "<span class='warning'>You are not the high priest, and therefore cannot select a religious sect.</span>")
		return
	if(!user.canUseTopic(parent, BE_CLOSE, FALSE, NO_TK))
		to_chat(user, "<span class='warning'>You cannot select a sect at this time.</span>")
		return
	var/mob/living/carbon/human/testsubject
	var/datum/religion_sect/pathed = path
	if(user == testsubject)
		if(testsubject.dna.species.id == pathed.req_species)
			pathed.starter = TRUE
			pathed.sect_requirements = "Unlocked"//////////////////////////////// work on this path shit ugh
			to_chat(user,"<span class='big notice'>Congratulations, you've unlocked [pathed.name]!</span>")
	if(!pathed.starter)
		to_chat(user, "<span class='warning'>You cannot select this sect until you complete its requirements.</span>")
		return
	if(GLOB.religious_sect)
		return
	GLOB.religious_sect = new path()
	for(var/i in GLOB.player_list)
		if(!isliving(i))
			continue
		var/mob/living/am_i_holy_living = i
		if(!am_i_holy_living.mind?.holy_role)
			continue
		GLOB.religious_sect.on_conversion(am_i_holy_living)
	easy_access_sect = GLOB.religious_sect
	after_sect_select_cb.Invoke()```
However I'm running into runtime issues where This or similar happens
![image](https://github.com/BeeStation/BeeStation-Hornet/assets/17776299/6e63ff84-3720-4930-913e-2310cfa3dafc)
 and I've been trying to sortit for about 4 hours now lmao

**Comment by ro5490** [comment]
// Need to fix religious_tool.dm between line 122 and 154 - gotta figure out a fix for that runtime

**Comment by EvilDragonfiend** [comment]
oh, actually I misunderstood the sect glob thing.
here's instruction.

1. declare `GLOBAL_LIST_EMPTY(religion_sect_datums)` in somwher. preferably in religion tool dm.
2. go to `global_lists.dm`, and do the thing here
```dm
/proc/init_religion_sects()
	for(var/path in subtypesof(/datum/religion_sect))
		var/datum/religion_sect/each_sect = new path()
		GLOB.religion_sect_datums += each_sect
```

**Comment by EvilDragonfiend** [comment]
so, `for(var/path in GLOB.religious_sect)` was a wrong suggestion. It had to be a list that's initialized from somehow.

File: code/datums/components/religious_tool.dm [hunk 1]
```diff
@@ -124,11 +125,21 @@
...
+	var/mob/living/carbon/human/testsubject
+	var/datum/religion_sect/pathed = path
+	if(user == testsubject)
+		if(testsubject.dna.species.id == pathed.req_species)
+			pathed.starter = TRUE
+			pathed.sect_requirements = "Unlocked"//////////////////////////////// work on this path shit ugh
+			to_chat(user,"<span class='big notice'>Congratulations, you've unlocked [pathed.name]!</span>")
+	if(!pathed.starter)
+		to_chat(user, "<span class='warning'>You cannot select this sect until you complete its requirements.</span>")
+		return
```
Comment (EvilDragonfiend) [inline code/datums/components/religious_tool.dm]: no, no, you can handle that in `generate_available_sects()` (but I suggested something wrong before.) Revoke this part fully.

File: code/modules/religion/religion_sects.dm [hunk 2]
```diff
@@ -22,6 +22,10 @@
 	var/starter = TRUE
 	/// species traits that block you from picking
 	var/invalidating_qualities = NONE
+	/// Sect Unlock Requirements
+	var/sect_requirements = "No unlock requirements"
+	/// Sect Species Requirements (Use species.id)
+	var/req_species = "Unlocked"
```
Comment (EvilDragonfiend) [inline code/modules/religion/religion_sects.dm]: We don't need this. remove these.

File: code/modules/religion/religion_sects.dm [hunk 3]
```diff
@@ -46,14 +50,14 @@
...
+/datum/religion_sect/New(atom/religious_tool, mob/living/creator)
 	. = ..()
 	if(desired_items)
 		desired_items_typecache = typecacheof(desired_items)
-	on_select()
+	on_select(religious_tool, creator)
 
 /// Activates once selected
-/datum/religion_sect/proc/on_select()
+/datum/religion_sect/proc/on_select(atom/religious_tool, mob/living/user)
```
Comment (EvilDragonfiend) [inline code/modules/religion/religion_sects.dm]: Revert these changes.

File: code/datums/components/religious_tool.dm [hunk 4]
```diff
@@ -186,6 +197,7 @@
 		var/datum/religion_sect/not_a_real_instance_rs = path
 		sect["name"] = initial(not_a_real_instance_rs.name)
 		sect["desc"] = initial(not_a_real_instance_rs.desc)
+		sect["req"] = initial(not_a_real_instance_rs.sect_requirements)
```
Comment (EvilDragonfiend) [inline code/datums/components/religious_tool.dm]: remove this

File: code/datums/components/religious_tool.dm [hunk 5]
```diff
@@ -94,6 +94,7 @@
 		data["name"] = easy_access_sect.name
 		data["desc"] = easy_access_sect.desc
 		data["quote"] = easy_access_sect.quote
+		data["req"] = easy_access_sect.sect_requirements
```
Comment (EvilDragonfiend) [inline code/datums/components/religious_tool.dm]: remove this

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]
Don't forget the solution I wrote above. `generate_available_sects()` needs a new global list datum for sects.

**Comment by EvilDragonfiend** [comment]
Feel free to ping in discord if you still have a trouble

**Comment by ro5490** [comment]
Still unsure what I'm meant to be doing with the new Global list, and  `generate_available_sects()` 
I've added the `/is_available` proc to `/datum/religion_sect`, removed the things you mentioned, and added the `is_available` to the shadow sect
where/how should I be referencing the new Global list, because as far as I could tell from testing, `generate_available_sects` doesnt get called because `easy_access_sect` gets set and used instead

**Comment by ro5490** [comment]
tried 
```
        var/datum/religion_sect/checks = path
        if(checks.is_available())
```
which triggered 
```
/datum/component/religious_tool/proc/perform_rite(mob/living/user, path)
    if(!ispath(text2path(path), /datum/religion_rites))
        message_admins("[ADMIN_LOOKUPFLW(usr)] has tried to spawn an item when performing a rite.")
        return
```
Also, the list seems to default to showing the debug/base sect

File: code/datums/components/religious_tool.dm [hunk 6]
```diff
@@ -177,11 +177,12 @@
 	var/list/sects_to_pick = list()
 	var/human_highpriest = ishuman(user)
 	var/mob/living/carbon/human/highpriest = user
-	for(var/path in subtypesof(/datum/religion_sect))
-		if(human_highpriest && initial(easy_access_sect.invalidating_qualities))
-			var/datum/species/highpriest_species = highpriest.dna.species
-			if(initial(easy_access_sect.invalidating_qualities) in highpriest_species.inherent_traits)
-				continue
+	for(var/path in GLOB.religion_sect_datums)
```
Comment (EvilDragonfiend) [inline code/datums/components/religious_tool.dm]:
```suggestion
	for(var/datum/religion_sect/each_sect in GLOB.religion_sect_datums)
		var/path = each_sect.type
```

File: code/datums/components/religious_tool.dm [hunk 7]
```diff
@@ -177,11 +177,12 @@
...
+		if(easy_access_sect.is_available())
```
Comment (EvilDragonfiend) [inline code/datums/components/religious_tool.dm]:
```suggestion
		if(!each_sect.is_available())
			continue
```

File: code/datums/components/religious_tool.dm [hunk 8]
```diff
@@ -177,11 +177,12 @@
...
+			if(human_highpriest && initial(easy_access_sect.invalidating_qualities))
+				var/datum/species/highpriest_species = highpriest.dna.species
+				if(initial(easy_access_sect.invalidating_qualities) in highpriest_species.inherent_traits)
+					continue
```
Comment (EvilDragonfiend) [inline code/datums/components/religious_tool.dm]: You don't need this.
point of `is_available()` is doing this for each sect.

File: code/__HELPERS/global_lists.dm [hunk 9]
```diff
@@ -73,6 +73,8 @@
 
 	init_crafting_recipes(GLOB.crafting_recipes)
 
+	init_religion_sects(GLOB.religion_sect_datums)
```
Comment (EvilDragonfiend) [inline code/__HELPERS/global_lists.dm]:
```suggestion
	init_religion_sects()
```
You don't need to send an augment.

**Review by EvilDragonfiend — DISMISSED** [review]

**Comment by EvilDragonfiend** [comment]
Since the sect is now only limited to specific cases, I lift my key.

**Comment by ro5490** [comment]
All set and ready for review(s) and hopefully approval before halloween

**Comment by EvilDragonfiend** [comment]
Since I am heavily involved in the PR at this point, I'll avoid approving this to avoid a possible misunderstanding that It doesn't need any further review.

**Comment by ro5490** [comment]
Time to fish for reviewers and someone to merge ;-;
Hoping this can get added before the halloween species get taken out again

**Review by Rukofamicom — APPROVED** [review]
Not my area of expertise, but code seems fine to me. 

Given the heavy restriction of access (outside of halloween anyway) I think this is fine to merge as-is with the potential for tweaks later if something overpowered comes out of this.

**Pull request merged**